corehq/apps/saved_reports/tasks.py

```diff
@@ -62,26 +63,13 @@ def send_report_throttled(notification_id):
     run_every=crontab(hour="*", minute="*/15", day_of_week="*"),
     queue=getattr(settings, 'CELERY_PERIODIC_QUEUE', 'celery'),
 )
-def daily_reports():
-    for report_id in get_scheduled_report_ids('daily'):
-        send_delayed_report(report_id)
-
-
-@periodic_task(
-    run_every=crontab(hour="*", minute="*/15", day_of_week="*"),
-    queue=getattr(settings, 'CELERY_PERIODIC_QUEUE', 'celery'),
-)
-def weekly_reports():
-    for report_id in get_scheduled_report_ids('weekly'):
-        send_delayed_report(report_id)
+def initiate_queue_scheduled_reports():
```

Comment:
This pattern seems a little silly/wasteful to me, but I saw it used elsewhere so I thought I'd start with it.

Essentially the lifecycle for firing scheduled reports would now be

```
beat => celery_periodic => initiate_queue_scheduled_reports() => celery_periodic => queue_scheduled_reports() (serial task)
```
followed by the rest of the existing process
```
... => background_queue (one per scheduled report) => send_report(notification_id)
```
I think the ideal thing were if it were possible to call `queue_scheduled_reports()` directly instead of `queue_scheduled_reports.delay()`, but I'm not sure whether that works. Anyone know?

My thinking about why this might matter is that if we have a backlog in the `celery_periodic` queue and the time-to-start for `celery_periodic` is 10 minutes, then with the double round trip it'll take 20 minutes before the actual send_report tasks are even queued. If no one has suggestions, I'm fine not optimizing around this in the first pass.

Reply:
I don't see any reason why that wouldn't work. Agree that the way it is now seems odd.

Reply:
OK I made a note to myself to revisit this after the feature's rolled out.

Reply:
@czue are you suggesting that

```
@serial_task(...)
@periodic_task(...)
def queue_scheduled_reports():
    ...
```
would work? That doesn't seem right to me because both of those are a kind of `@task` decorator, so it'd be declared as a task in two different ways, and I don't have a mental model for how that'd work; I'm not even sure the behavior there is well defined.

What I was suggesting (and maybe what you were saying too) is to replace

```
@periodic_task(...)
def initiate_queue_scheduled_reports():
    queue_scheduled_reports.delay()
```
with
```
@periodic_task(...)
def initiate_queue_scheduled_reports():
    queue_scheduled_reports()  # no .delay here
```

What I would _want_ that to do is run `queue_scheduled_reports` directly without putting it back in the queue, but if serial_task finds another process is running `queue_scheduled_reports` at the same time, that it _do_ get put on the queue to be retried at a future date. This may be the actual behavior, but I thin my uncertainty comes from not knowing what `self.retry` does in the situation when a task function was called directly instead of through celery. It seems reasonable that what I want would happen there. It seems like a simple-enough thing to try, so maybe I'll just do a local test and see what happens.

Reply:
Actually just to amend what I said, I'll test that the main case works, as a sanity check. There's no reason the locking should work any different, so two running at a time isn't a concern, and if retry doesn't work the worst that'll happen is that we get a celery error and we wait until the next 15-minute mark to fire again. That's a remarkably inconsequential downside, so I'll keep an eye out for such a message in sentry once we deploy, but I won't try to trigger that locally to see what happens.

Reply:
Right, I see what you're saying in the first comment. And in the second you're just saying that making it `serial` isn't actually necessary because of locking, yeah? That also makes sense to me.

I think if for whatever reason we wanted to bring back the serial-ness granted by the `@serial_task` decorator, we could split it up into a `@serial` (that deals with locking and such) and `@task` (that actually makes it a task)? Basically extracting the contents of `[_inner](https://github.com/dimagi/commcare-hq/blob/master/corehq/util/decorators.py#L121-L135)` and making them a separate decorator and then having `@serial_task` and this code use that. 

But the above doesn't seem necessary in this case since as you pointed out we're already managing locking via the DB.

Reply:
Oh I wasn't even factoring in db level locking, which I'm not super confident about/haven't thought through. I was referring to what I eventually did in this commit: https://github.com/dimagi/commcare-hq/pull/23737/commits/36b9e3d8a4398df8245ed91f86e59c7f70f90889, i.e.

```
@periodic_task(...)
def initiate_queue_scheduled_reports():
    queue_scheduled_reports()  # no .delay here
```

`queue_scheduled_reports` is still a `@serial_task`. My [second comment](https://github.com/dimagi/commcare-hq/pull/23737#discussion_r270539567) was really just saying that I realized the details of how it works in the retry case (when it can't acquire the redis lock) weren't actually that important, since we don't really depend on the retry anyway.

What you said about splitting `@serial_task` into `@serial` and the existing `@task` makes sense, though I think you'd have to be tricky to get the same retry behavior for a serial task in that case.

Anyways, I think I've made the change I wanted to make that came out of this convo.

Reply:
> Anyways, I think I've made the change I wanted to make that came out of this convo.

 Sounds good. Sorry if I confused things...